src/GafferML/ImageToTensor.cpp

```diff
@@ -0,0 +1,294 @@
+			[&] ( const ImagePlug *image, const string &channelName, const Imath::V2i &tileOrigin, const IECore::MurmurHash &tileHash )
+			{
+				h.append( tileHash );
+			},
+			dataWindow,
+			ImageAlgo::TopToBottom
+		);
+
+		h.append( dataWindow );
+	}
+	else
+	{
+		ComputeNode::hash( output, context, h );
+	}
+}
+
+void ImageToTensor::compute( Gaffer::ValuePlug *output, const Gaffer::Context *context ) const
+{
+	if( output == tensorPlug() )
+	{
+		ConstStringVectorDataPtr channelsData = channelsPlug()->getValue();
+		const auto &channels = channelsData->readable();
+		const bool interleaveChannels = interleaveChannelsPlug()->getValue();
+
+		ImagePlug::ViewScope viewScope( context );
+		const std::string view = viewPlug()->getValue();
+		viewScope.setViewNameChecked( &view, imagePlug()->viewNames().get() );
+
+		const Box2i dataWindow = imagePlug()->dataWindow();
+		ConstStringVectorDataPtr inChannels = imagePlug()->channelNamesPlug()->getValue();
+		const size_t numPixels = dataWindow.size().x * dataWindow.size().y;
+
+		FloatVectorDataPtr bufferData = new FloatVectorData;
+		vector<float> &buffer = bufferData->writable();
+		buffer.resize( numPixels * channels.size() );
+
+		boost::container::flat_map<std::string, size_t> channelIndices;
+		for( size_t i = 0; i < channels.size(); ++i )
+		{
+			channelIndices[channels[i]] = i;
+		}
+
+		ImageAlgo::parallelProcessTiles(
+			imagePlug(),
+			channels,
+			[&] ( const ImagePlug *image, const string &channelName, const Imath::V2i &tileOrigin )
+			{
+				IECore::Canceller::check( context->canceller() );
+
+				if( !ImageAlgo::channelExists( inChannels->readable(), channelName ) )
+				{
+					throw IECore::Exception( fmt::format( "Channel \"{}\" does not exist", channelName ) );
+				}
+
+				ConstFloatVectorDataPtr channelData = image->channelDataPlug()->getValue();
+				const Box2i tileBound( tileOrigin, tileOrigin + V2i( ImagePlug::tileSize() ) );
+				const Box2i validTileBound = BufferAlgo::intersection( tileBound, dataWindow );
+
+				const size_t channelIndex = channelIndices[channelName];
+				float *dstData = buffer.data();
+				size_t dstStride;
+				if( interleaveChannels )
+				{
+					dstData += channelIndex;
+					dstStride = channels.size();
+				}
+				else
+				{
+					dstData += numPixels * channelIndex;
+					dstStride = 1;
+				}
+
+				const float *sourceData = channelData->readable().data();
+
+				for( V2i p = validTileBound.min; p.y < validTileBound.max.y; ++p.y )
+				{
+					size_t dstIndex = BufferAlgo::index( V2i( p.x, dataWindow.max.y - p.y - 1 ), dataWindow ) * dstStride;
+					size_t srcIndex = BufferAlgo::index( p, tileBound );
+					for( int x = validTileBound.min.x; x < validTileBound.max.x; ++x )
+					{
+						dstData[dstIndex] = sourceData[srcIndex++];
+						dstIndex += dstStride;
+					}
+				}
+			}
+		);
+
+		vector<int64_t> shape;
+		if( interleaveChannels )
+		{
+			shape = { 1, dataWindow.size().y, dataWindow.size().x, (int64_t)channels.size() };
```

Comment:
Yeah, the compiler complains about a narrowing conversion otherwise, but in practice `channels.size()` will never be big enough for narrowing to be an issue.

Reply:
Makes sense